src/v/cluster/partition_balancer_planner.cc

```diff
@@ -0,0 +1,420 @@
+              a.replicas, rrs, SIZE_MAX);
+            if (available_replicas.size() != a.replicas.size()) {
+                auto partition_size = get_partition_size(ntp, rrs);
+                if (
+                  !partition_size.has_value()
+                  || !is_partition_movement_possible(a.replicas, rrs)) {
+                    continue;
+                }
+                auto new_allocation_units = get_reallocation(
+                  a, t.second, partition_size.value(), false, rrs);
+                if (!new_allocation_units) {
+                    vlog(
+                      clusterlog.info,
+                      "Can't reallocate {} from down node",
+                      ntp);
+                } else {
+                    reassignments.emplace_back(ntp_reassignments{
+                      .ntp = ntp,
+                      .allocation_units = std::move(
+                        new_allocation_units.value())});
+                }
+            }
+
+            // End adding movements if batch is collected
+            if (
+              rrs.planned_movement_disk_size
+              >= _config.movement_disk_size_batch) {
+                return;
+            }
+        }
+    }
+}
+
+/*
+ * Function is trying to move ntps out of node that are violating
+ * max_disk_usage_ratio It takes nodes in free space rate order For each node it
+ * is trying to collect set of partitions to move Partitions are selected in
+ * ascending order of their size.
+ */
+void partition_balancer_planner::get_full_node_reassignments(
+  std::vector<ntp_reassignments>& reassignments,
+  reallocation_request_state& rrs) {
+    absl::flat_hash_map<model::node_id, std::vector<model::ntp>> ntp_on_nodes;
+    for (const auto& t : _topic_table.topics_map()) {
+        for (const auto& a : t.second.get_assignments()) {
+            for (const auto& r : a.replicas) {
+                ntp_on_nodes[r.node_id].emplace_back(
+                  t.first.ns, t.first.tp, a.id);
+            }
+        }
+    }
+
+    for (const auto& node_disk_space : rrs.full_nodes_disk_sizes) {
+        if (
+          rrs.planned_movement_disk_size >= _config.movement_disk_size_batch) {
+            return;
+        }
+        absl::btree_multimap<size_t, model::ntp> ntp_on_node_sizes;
+        for (const auto& ntp : ntp_on_nodes[node_disk_space.node_id]) {
+            auto partition_size_opt = get_partition_size(ntp, rrs);
+            if (partition_size_opt.has_value()) {
+                ntp_on_node_sizes.emplace(partition_size_opt.value(), ntp);
+            }
+        }
+        auto ntp_size_it = ntp_on_node_sizes.begin();
+        while (double(
+                 rrs.node_released_disk_size[node_disk_space.node_id]
+                 + node_disk_space.free_space)
+                   / double(node_disk_space.total_space)
+                 < double(1) - _config.max_disk_usage_ratio
+               && ntp_size_it != ntp_on_node_sizes.end()) {
```

Comment:
this conditional would be nice to factor out into a lambda or something so that it's easier to read.